docs/custom-plugins/visitor.md

```diff
@@ -0,0 +1,113 @@
+# Visitor pattern
+
+_Visitor_ is a design pattern that allows operations to be performed on individual elements in a more complex structure. Redocly uses it as the basis for custom plugins where every node in a document must be evaluated, for example when applying rules or decorators.
+
+To understand how this applies to your API description, think of the document as a tree structure. The top level elements are entries like `info` and `components`. To examine the `description` field in the `info` section, the visitor goes to the `info` node first, then on to the `description` node. This pattern is repeated all over the document as the visitor pattern works its way around the document tree.
+
+<a id="visitor-object"></a><a id="format-of-visitor-object"></a>
+
+## Structure of the visitor object
+
+In your plugin, create a JavaScript visitor object, and describe the functionality required for each type of node.
+
+Redocly CLI calls `enter()` while going down the tree and `leave()` going up the tree after processing the node and its children.
+`skip()` predicate is called and if it returns `true` the node is ignored for this visitor.
+
+
+```js
+function ExampleRule() {
+  const seen = {};
+  return {
+    Root: {
+      leave() {
+        // check something and report
+      }
+    }
+    Operation: {
+      enter(operation, ctx) {
+        seen[operation.operationId] = true;
+      },
+    }
+  };
+}
+```
+
+Keys of the object are one of the following:
+
+- document-specific node types, such as the [OpenAPI node types](../../openapi-visual-reference/openapi-node-types.md).
+- `any` - visitor is called on every node.
+- `ref` - visitor is called on $ref nodes.
+
+
+## Visitors execution and $ref
+
+Top level **visitor functions** run only once for each node.
+If the same node is referenced by the $ref multiple times,
+top-level **visitor functions** are executed only once for this node.
+
+This works fine for most context-free rules which check basic things. If you need contextual info you should use [nested visitors](#nested-visitors).
+
+## Nested visitors
+
+
+Visitor object (if it is not `any` or `ref`) can define [nested visitors](#nested-visitors).
+
+Here is a basic example of a nested visitor:
+
+```js
+function ExampleRule() {
+  const seen = {};
+  return {
+    Operation: {
+      // skip: (value, key) => ... // if needed
+      // enter(operation) {} // if needed
+      Schema(schema, ctx, parents) {
```

Comment:
Wouldn't it be better to always explicitly use the `enter` visitor in our documentation? I remember, named visitor functions confused me when I saw them first time. It still looks strange to me that the same thing could be an object or a function  😅

Reply:
Ah, good spot. I didn't update this bit (just moved it to a new file) but I agree an object would be clearer. I'll switch the example.

Reply:
@tatomyr can you take another look? I tried to replace this existing example with an equivalent. It seems to work the same but would be good to get your eyes on it too